Creating RPC API and base structure to handle native calls in the omni-executor

silva-fj opened the pull request:
This PR introduces the base structure to start migrating the core functionality of the ID worker (OmniAccount and VC), setting up the RPC api and the native task handler.

It also integrates some of the types and code from the legacy code. (via imports or copying) required to handle AesRequest and the interfaces for authentication.

silva-fj committed:
setting up rpc crate

silva-fj committed:
adding dependencies

silva-fj committed:
initial rpc server setup

silva-fj committed:
running rpc server

silva-fj committed:
adding dependencies to the workspace

silva-fj committed:
adding dependencies to rpc crate

silva-fj committed:
adding shielding_key type

silva-fj committed:
refactoring rpc and adding shielding_key init

silva-fj committed:
adding get_shielding_key method

silva-fj committed:
adding dependencies to the workspace

silva-fj committed:
adding utils

silva-fj committed:
adding dependencies to rpc-server

silva-fj committed:
adding crypto module and refactoring shielding_key

silva-fj committed:
improving get_shielding_key

silva-fj committed:
fixing imports

silva-fj committed:
adding aes256

silva-fj committed:
updating server imports

silva-fj committed:
adding submit_native_request method (WIP)

silva-fj committed:
fixing rsa dependencies

silva-fj committed:
extracting out crypto

silva-fj committed:
small refactoring, updating usage of crypto functions

silva-fj committed:
implementing Display for hex error

silva-fj committed:
native-call-executor initial setup

silva-fj committed:
extending rpc context

silva-fj committed:
error_code initial setup

silva-fj committed:
adding submit_native_aes_request (WIP)

silva-fj committed:
running native-call-executor and fixing context

silva-fj committed:
updating get_shielding_key method, adding error code response

silva-fj committed:
updating deps

silva-fj committed:
updating rustls-webpki to the latest version to fix compatilibity issue

silva-fj committed:
updating lock files

silva-fj committed:
installing core primitives to the workspace

silva-fj committed:
adding sender_identity to native calls

silva-fj committed:
small cleanup

silva-fj committed:
cleaning up dependencies

silva-fj committed:
fixing fmt issues

silva-fj committed:
updating method name

silva-fj committed:
small refactoring

silva-fj committed:
adding secp256k1 crypto module

silva-fj committed:
adding more dependencies to the workspace

silva-fj committed:
exposing sp-core crypto modules

silva-fj committed:
setting up heima-primitives crate

silva-fj committed:
Merge remote-tracking branch 'origin/dev' into p-1252-create-an-rpc-api-and-migrate-relevant-methods

silva-fj committed:
reverting commit

silva-fj committed:
adding signature module to parentchain-primitives

kziemianek commented on tee-worker/omni-executor/rpc-server/src/methods/submit_native_aes_request.rs:
```diff
@@ -0,0 +1,122 @@
+	ctx: Arc<RpcContext>,
+) -> Result<NativeCall, ErrorObject<'a>> {
+	#[cfg(feature = "gramine-quote")]
+	if request.mrenclave != ctx.mrenclave {
```
can we use `[0; 32]` for environment where real mrenclave is not present ?

silva-fj replied:
Yes, it makes sense. This method does not need to know about gramine 👍🏼

silva-fj committed:
small improvement to submit_native_aes_request method

silva-fj committed:
adding heima-primitives to the workspace

silva-fj committed:
updating Cargo.lock file

silva-fj committed:
importing real Identity type

silva-fj committed:
adding heima prefix

silva-fj committed:
adding more dependencies to the workspace

silva-fj committed:
exposing more crypto primitives

silva-fj committed:
implementing HeimaMultiSignature

silva-fj committed:
exposing heima-primitives from parentchain-primitives

silva-fj committed:
updating RpcContext

silva-fj committed:
adding request module

silva-fj committed:
adding parentchain-primitives dependency to rpc-server

silva-fj committed:
adding signature_message_prefix to NativeCall

silva-fj committed:
adding authentication module (WIP)

silva-fj committed:
updating error_code

silva-fj committed:
refining names

silva-fj committed:
refactoring submit_aes_request method

silva-fj committed:
fixing fmt issues

silva-fj committed:
updating comment

silva-fj committed:
refactoring names

silva-fj committed:
removing unused method for now

silva-fj committed:
moving storage to general local crates

silva-fj committed:
adding skeleton for google oauth2 authentication

silva-fj committed:
keeping control of what is exported from common core-primitives

silva-fj committed:
exposing more primitive types

silva-fj committed:
refactoring native-task-executor into native-task-handler

silva-fj committed:
fixing usages of the native-task-handler

silva-fj committed:
refactoring submit_aes_request

silva-fj committed:
ordering imports

silva-fj committed:
moving native_call to executor-core

silva-fj committed:
moving GetEventId trait to parentchain primitives

silva-fj committed:
updating rpc-server

silva-fj committed:
updating native-task-handler

silva-fj committed:
updating Cargo.lock file

silva-fj committed:
refactor(deps): alphabetize workspace dependencies

Reorder dependencies in omni-executor/Cargo.toml alphabetically for better
readability and maintainability. This includes both external crates and
local workspace dependencies.

silva-fj committed:
fixing fmt issue

silva-fj committed:
updating webpki on the parachain as well

silva-fj committed:
Revert "updating rustls-webpki to the latest version to fix compatilibity issue"

This reverts commit 4e7d833787d083afd61450537822f421155e752b.

silva-fj committed:
Revert "updating webpki on the parachain as well"

This reverts commit cf4823616949751bd3c90d32b463172034b4106e.

silva-fj committed:
Revert "updating lock files"

This reverts commit 3097a48bf408fba897a021a4278f66db174bd891.

silva-fj committed:
fix compilation; hide webpki behind a feature

silva-fj committed:
Merge branch 'dev' into p-1252-create-an-rpc-api-and-migrate-relevant-methods

silva-fj committed:
removing unused code

silva-fj committed:
fixing clippy issues

silva-fj committed:
fixing tests

silva-fj committed:
updating docker build

silva-fj committed:
fixing typo

silva-fj committed:
updating docker context on CI

Kailai-Wang commented on common/primitives/core/Cargo.toml:
```diff
@@ -17,7 +17,7 @@ serde = { version = "1.0", default-features = false, features = ["derive", "allo
 strum = { version = "0.26", default-features = false }
 strum_macros = { version = "0.26", default-features = false }
-webpki = { version = "=0.102.0-alpha.3", git = "https://github.com/rustls/webpki", rev = "da923ed", package = "rustls-webpki", default-features = false, features = ["alloc", "ring"] }
+webpki = { version = "=0.102.0-alpha.3", git = "https://github.com/rustls/webpki", rev = "da923ed", package = "rustls-webpki", default-features = false, features = ["alloc", "ring"], optional = true }
```
Why this? Is it because worker needs teebag primitives but don't want to introduce webpki? Or any dep conflict with this crate? (I remember we need to pin this version for no-std support)

silva-fj replied:
Yes, there are some conflicts with this crate so I made it optional to be able to use the `core-primitives` in the worker

Kailai-Wang commented on tee-worker/omni-executor/crypto/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "crypto"
```
I think this crate and the `storage` crate below need a new name, otherwise it's too general

Common paradigm is using project as prefix, like `sp-`, `subxt-` etc

Shall we call it `hm-crypto` (heima-crypto) if it's a crate that can be used by repos, or `hoa-crypto` (heima-omni-account crypto) if it's omni-account specific?

silva-fj replied:
Adding a prefix sounds good. I was not happy with the names either 👍🏼

Kailai-Wang commented on tee-worker/omni-executor/parentchain/primitives/src/signature/bitcoin_signature.rs:
I think those files don't need to end with `_signature.rs` as they are under `signature/` already

Kailai-Wang commented on tee-worker/omni-executor/rpc-server/src/methods/submit_aes_request.rs:
```diff
@@ -0,0 +1,115 @@
+
+pub fn register_submit_aes_request(module: &mut RpcModule<RpcContext>) {
+	module
+		.register_async_method("native_submitAesRequest", |params, ctx, _| async move {
```
Would you add another `native_submitPlainRequest` which doesn't need to be encrypted with aes?

Also a question:

rpc methods start with `native_` stand for a nativeCall request, right? A `NativeCall` isn't equal to `AuthenticatedCall`?

Like we should differentiate between aes-encrypted `nativeCall`, and authenticated `nativeCall`, and even aes-encrypted authenticated `nativeCall`:
- aes means we want the payload to be encrypted/private
- authenticated means we want to verify the origin of the requester

I'm wondering if there're better structure/names to make these more readable

silva-fj replied:
The `AuthenticatedCall` is the equivalent of `TrustedCallSigned` or `TrustedCallAuthenticated` in the identity worker and `NativeCall` is the equivalent of `TrustedCall`. 

How about if we call them `NativeCallAuthenticated` and `NativeCall`? 

Now that you mentioned this about the encrypted requests, why do we need them for direct calls? Would `native_submitNativeCallAuthenticated` be enough for these calls? The client would only need to send the `AuthenticatedNativeCall` encoded.

Also, here we are skipping another wrapper (`TrustedOperation`). If it's OK to remove the need for encrypting the requests for direct calls it would simplify even more the client side. What do you think @Kailai-Wang ?

Kailai-Wang reviewed (approved):
The skeleton looks good, thanks!

silva-fj committed:
Merge remote-tracking branch 'origin/dev' into p-1252-create-an-rpc-api-and-migrate-relevant-methods

silva-fj committed:
small refactoring

silva-fj committed:
renaming crypto to executor-crypto

silva-fj committed:
renaming storage to executor-storage

silva-fj committed:
fixing fmt issue

silva-fj commented:
I'll wait for https://github.com/litentry/heima/pull/3235 and https://github.com/litentry/heima/pull/3238 reviews to get them merged here first to avoid merging conflicts.

silva-fj committed:
Adding support for email authentication in the omni-executor (#3235)

* adding heima-primitives to storage

* adding verification_code storage

* adding new mailer crate

* implementing mailer service

* using error code for parsing errors

* exposing sp_core's AccountId from parentchain-primitives

* updating verification_code storage

* using AccountId from parentchain-primitives

* adding OmniAccount's member_account_hash storage

* updating Cargo.lock file

* moving primitives to its own crate, keeping all types in one place

* adding trait to convert types from subxt to primitives

* initializing omni_account storages

* refactoring trait name

* refactoring member_account storage

* injecting MemberAccountStorage to the event_handler

* removing comments

* updating MemberAccountStorage when handling events

* updating gitignore

* refactoring storage, using better names

* refactoring storage, using a single instance of rocksdb

* setting up haima-identity-verification crate

* adding helpers

* setting up email identity-verification

* cleaning up mailer

* refactoring mailer exports

* injecting mailer to rpc server context

* adding request_email_verification_code method

* moving Storage trait to storage crate

* implementing verify_email_authentication

* refactoring: using handle_request with the blocking api

* removing unused imports

* fixing fmt issue

* adding comments

* removing verification code after using it

* Adding support for AuthToken authentication in the omni-executor (#3238)

* adding jsonwebtoken dependency

* adding jwt module to crypto

* adding hex_display utils

* moving hex utils to primitives

* improving jwt

* refactoring rpc client

* adding more dependencies to rpc context

* implementing verify_auth_token_authentication

* refactoring jwt

* adding heima-authentication

* refactoring auth_token validation

* refactoring rpc context

* updating parentchain-api-interface

* using block number type from primitives

* remove extra spacing

* cleaning up utils

* refactoring block number type to be compatible with the parachain

* exposing subxt AccountId from rpc-client

* Revert "exposing subxt AccountId from rpc-client"

This reverts commit 02d36c1221474d7ec4649053ff0d3237eb7e8282.

* adding TaskHandlerContext

* renaming jwt functions

* adding native-task-handler result types

* adding request_auth_token native call

* adding dependencies

* adding error variant

* initial implementation of request_auth_token native call handler

* refactoring handler name

* passing client-factory instead of client to the contexts

* small improvement

* refactoring rpc server context

* refactoring native call handler context

* extracting parentchain signer to its own crate

* moving metadata to parentchain rpc-client

* exporting types from rpc-client

* adding internal error variant

* adding TransactionSigner to the native-task-handler context

* sending tx to parentchain

* creating type alias

* injecting tx signer in the listener

* moving get_signer function to parentchain-signer crate

silva-fj committed:
Merge branch 'dev' into p-1252-create-an-rpc-api-and-migrate-relevant-methods
